proxmox_virtual_environment_file broken

**Describe the bug**
An attempt to create proxmox_virtual_environment_file resource, specifically

```
resource "proxmox_virtual_environment_file" "debian_cloud_image" {
  lifecycle {
    ignore_changes = [
      source_file
    ]
  }

  content_type = "iso"
  datastore_id = "local"
  node_name    = pve_node

  source_file {
    path = "./debian-11-generic-amd64.img"
  }
}
```
ends up with crashed plugin when using provider newer than v0.28.0. I believe this commit brokes it:
https://github.com/bpg/terraform-provider-proxmox/commit/f901e711dd4e8cd59b3a1e34c58a1a03564bd13a


**To Reproduce**
Steps to reproduce the behavior:
1. Create a resource as above
2. Run terraform apply
3. See error

Please also provide a minimal Terraform configuration that reproduces the issue.

**Expected behavior**
Resource would be created :)

Working example, v0.28.0:

```
proxmox_virtual_environment_file.debian_cloud_image["dev-pve04"]: Creation complete after 2m36s [id=local:iso/debian-11-generic-amd64.img]
2023-08-23T20:15:16.517+0200 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2023-08-23T20:15:16.551+0200 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/bpg/proxmox/0.28.0/linux_amd64/terraform-provider-proxmox_v0.28.0 pid=1023245
2023-08-23T20:15:16.552+0200 [DEBUG] provider: plugin exited
2023-08-23T20:15:16.565+0200 [DEBUG] POST <redacted>
╷
│ Warning: Applied changes may be incomplete
│
│ The plan was created with the -target option in effect, so some changes requested in the configuration may have been ignored and the output values may not be fully updated. Run the following command to
│ verify that no other changes are pending:
│     terraform plan
│
│ Note that the -target option is not suitable for routine use, and is provided only for exceptional situations such as recovering from errors or mistakes, or when Terraform specifically suggests to use it
│ as part of an error message.
╵
2023-08-23T20:15:16.754+0200 [DEBUG] DELETE <redacted>/lock

Apply complete! Resources: 2 added, 0 changed, 1 destroyed.
```

broken, v0.30.1:
```
proxmox_virtual_environment_file.debian_cloud_image["dev-pve05"]: Still creating... [2m31s elapsed]
2023-08-23T20:18:45.344+0200 [DEBUG] provider.terraform-provider-proxmox_v0.30.1: Received HTTP Response: tf_http_trans_id=658488ec-955b-c1af-495a-c4f49c763f97 tf_provider_addr=registry.terraform.io/bpg/proxmox tf_resource_type=proxmox_virtual_environment_file Cache-Control=max-age=0 tf_http_op_type=response tf_http_res_body={"data":"UPID:dev-pve01:003F9B4A:06C7022D:64E64D85:imgcopy::root@pam:"} tf_http_res_status_code=200 Expires="Wed, 23 Aug 2023 18:18:45 GMT" tf_mux_provider=tf5to6server.v5tov6Server Content-Length=71 Content-Type=application/json;charset=UTF-8 @caller=github.com/hashicorp/terraform-plugin-sdk/v2@v2.27.0/helper/logging/logging_http_transport.go:162 Server=pve-api-daemon/3.0 tf_http_res_version=HTTP/1.1 tf_req_id=a51fd800-dd90-b937-d94e-603c03e133e5 tf_rpc=ApplyResourceChange @module=proxmox Date="Wed, 23 Aug 2023 18:18:45 GMT" Pragma=no-cache tf_http_res_status_reason="200 OK" timestamp=2023-08-23T20:18:45.344+0200
2023-08-23T20:18:45.345+0200 [DEBUG] provider.terraform-provider-proxmox_v0.30.1: Sending HTTP Request: tf_http_req_uri=/api2/json/nodes/dev-pve05/tasks/UPID:dev-pve01:003F9B4A:06C7022D:64E64D85:imgcopy::root@pam:/status @module=proxmox Cookie=PVEAuthCookie=PVE:root@pam:<redacted> tf_http_op_type=request tf_http_req_method=GET Accept=application/json Accept-Encoding=gzip tf_http_req_body= tf_http_trans_id=425a808a-2f8c-13a4-c579-2d0a4947a963 tf_rpc=ApplyResourceChange Host=10.227.0.11:8006 tf_mux_provider=tf5to6server.v5tov6Server tf_req_id=a51fd800-dd90-b937-d94e-603c03e133e5 tf_resource_type=proxmox_virtual_environment_file @caller=github.com/hashicorp/terraform-plugin-sdk/v2@v2.27.0/helper/logging/logging_http_transport.go:162 User-Agent=Go-http-client/1.1 tf_http_req_version=HTTP/1.1 tf_provider_addr=registry.terraform.io/bpg/proxmox timestamp=2023-08-23T20:18:45.345+0200
2023-08-23T20:18:45.372+0200 [DEBUG] provider.terraform-provider-proxmox_v0.30.1: Received HTTP Response: @module=proxmox tf_http_res_version=HTTP/1.1 tf_provider_addr=registry.terraform.io/bpg/proxmox tf_resource_type=proxmox_virtual_environment_file tf_http_res_body="{"data":null,"errors":{"upid":"no such task"}}" tf_http_res_status_code=400 tf_http_res_status_reason="400 Parameter verification failed." Cache-Control=max-age=0 Content-Length=46 Content-Type=application/json;charset=UTF-8 Expires="Wed, 23 Aug 2023 18:18:45 GMT" tf_http_trans_id=425a808a-2f8c-13a4-c579-2d0a4947a963 tf_mux_provider=tf5to6server.v5tov6Server tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-sdk/v2@v2.27.0/helper/logging/logging_http_transport.go:162 Date="Wed, 23 Aug 2023 18:18:45 GMT" Pragma=no-cache Server=pve-api-daemon/3.0 tf_http_op_type=response tf_req_id=a51fd800-dd90-b937-d94e-603c03e133e5 timestamp=2023-08-23T20:18:45.372+0200
2023-08-23T20:18:45.372+0200 [ERROR] provider.terraform-provider-proxmox_v0.30.1: failed to close file reader: tf_mux_provider=tf5to6server.v5tov6Server tf_req_id=a51fd800-dd90-b937-d94e-603c03e133e5 tf_resource_type=proxmox_virtual_environment_file @caller=github.com/bpg/terraform-provider-proxmox/proxmox/nodes/storage.go:243 error="close /tmp/multipart3390830491: file already closed" tf_provider_addr=registry.terraform.io/bpg/proxmox tf_rpc=ApplyResourceChange @module=proxmox timestamp=2023-08-23T20:18:45.372+0200
2023-08-23T20:18:45.448+0200 [ERROR] provider.terraform-provider-proxmox_v0.30.1: Response contains error diagnostic: diagnostic_severity=ERROR tf_proto_version=6.3 tf_provider_addr=registry.terraform.io/bpg/proxmox tf_req_id=a51fd800-dd90-b937-d94e-603c03e133e5 tf_rpc=ApplyResourceChange @module=sdk.proto diagnostic_detail= diagnostic_summary="error uploading file to datastore local: failed waiting for upload - error retrievinf task status: received an HTTP 400 response - Reason: Parameter verification failed. (upid: no such task)" tf_resource_type=proxmox_virtual_environment_file @caller=github.com/hashicorp/terraform-plugin-go@v0.18.0/tfprotov6/internal/diag/diagnostics.go:58 timestamp=2023-08-23T20:18:45.448+0200
2023-08-23T20:18:45.449+0200 [ERROR] vertex "proxmox_virtual_environment_file.debian_cloud_image[\"dev-pve05\"]" error: error uploading file to datastore local: failed waiting for upload - error retrievinf task status: received an HTTP 400 response - Reason: Parameter verification failed. (upid: no such task)
2023-08-23T20:18:45.459+0200 [DEBUG] POST <redacted>
╷
│ Warning: Applied changes may be incomplete
│
│ The plan was created with the -target option in effect, so some changes requested in the configuration may have been ignored and the output values may not be fully updated. Run the following command to
│ verify that no other changes are pending:
│     terraform plan
│
│ Note that the -target option is not suitable for routine use, and is provided only for exceptional situations such as recovering from errors or mistakes, or when Terraform specifically suggests to use it
│ as part of an error message.
╵
╷
│ Error: error uploading file to datastore local: failed waiting for upload - error retrievinf task status: received an HTTP 400 response - Reason: Parameter verification failed. (upid: no such task)
│
│   with proxmox_virtual_environment_file.debian_cloud_image["dev-pve05"],
│   on debian-image.tf line 21, in resource "proxmox_virtual_environment_file" "debian_cloud_image":
│   21: resource "proxmox_virtual_environment_file" "debian_cloud_image" {
```

--- a/proxmox/nodes/tasks/client.go
+++ b/proxmox/nodes/tasks/client.go
@@ -8,6 +8,7 @@ package tasks
 
 import (
 	"fmt"
+	"net/url"
 
 	"github.com/bpg/terraform-provider-proxmox/proxmox/api"
 )
@@ -18,8 +19,17 @@ type Client struct {
 }
 
 // ExpandPath expands a path relative to the client's base path.
-func (c *Client) ExpandPath(path string) string {
-	return c.Client.ExpandPath(
-		fmt.Sprintf("tasks/%s", path),
-	)
+func (c *Client) ExpandPath(_ string) string {
+	panic("ExpandPath of tasks.Client must not be used. Use BuildPath instead.")
+}
+
+// BuildPath builds a path using information from Task ID.
+func (c *Client) BuildPath(taskID string, path string) (string, error) {
+	tid, err := ParseTaskID(taskID)
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("nodes/%s/tasks/%s/%s",
+		url.PathEscape(tid.NodeName), url.PathEscape(taskID), url.PathEscape(path)), nil
 }

--- a/proxmox/nodes/tasks/tasks.go
+++ b/proxmox/nodes/tasks/tasks.go
@@ -11,7 +11,6 @@ import (
 	"errors"
 	"fmt"
 	"net/http"
-	"net/url"
 	"time"
 
 	"github.com/bpg/terraform-provider-proxmox/proxmox/api"
@@ -21,10 +20,15 @@ import (
 func (c *Client) GetTaskStatus(ctx context.Context, upid string) (*GetTaskStatusResponseData, error) {
 	resBody := &GetTaskStatusResponseBody{}
 
-	err := c.DoRequest(
+	path, err := c.BuildPath(upid, "status")
+	if err != nil {
+		return nil, fmt.Errorf("error building path for task status: %w", err)
+	}
+
+	err = c.DoRequest(
 		ctx,
 		http.MethodGet,
-		c.ExpandPath(fmt.Sprintf("%s/status", url.PathEscape(upid))),
+		path,
 		nil,
 		resBody,
 	)

--- a/proxmox/nodes/tasks/tasks_types.go
+++ b/proxmox/nodes/tasks/tasks_types.go
@@ -78,7 +78,7 @@ func ParseTaskID(taskID string) (TaskID, error) {
 		NodeName:  parts[1],
 		PID:       int(pid),
 		PStart:    int(pstart),
-		StartTime: time.Unix(stime, 0),
+		StartTime: time.Unix(stime, 0).UTC(),
 		Type:      parts[5],
 		ID:        parts[6],
 		User:      parts[7],

--- a/proxmox/nodes/tasks/tasks_types.go
+++ b/proxmox/nodes/tasks/tasks_types.go
@@ -6,6 +6,13 @@
 
 package tasks
 
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+)
+
 // GetTaskStatusResponseBody contains the body from a node get task status response.
 type GetTaskStatusResponseBody struct {
 	Data *GetTaskStatusResponseData `json:"data,omitempty"`
@@ -16,4 +23,64 @@ type GetTaskStatusResponseData struct {
 	PID      int    `json:"pid,omitempty"`
 	Status   string `json:"status,omitempty"`
 	ExitCode string `json:"exitstatus,omitempty"`
+}
+
+// TaskID contains the components of a PVE task ID.
+type TaskID struct {
+	NodeName  string
+	PID       int
+	PStart    int
+	StartTime time.Time
+	Type      string
+	ID        string
+	User      string
+}
+
+// ParseTaskID parses a task ID into its component parts.
+// The task ID is expected to be in the format of:
+//
+//	UPID:<node_name>:<pid_in_hex>:<pstart_in_hex>:<starttime_in_hex>:<type>:<id (optional)>:<user>@<realm>:
+func ParseTaskID(taskID string) (TaskID, error) {
+	parts := strings.SplitN(taskID, ":", 9)
+
+	if parts[0] != "UPID" || len(parts) < 8 {
+		return TaskID{}, fmt.Errorf("invalid task ID format: %s", taskID)
+	}
+
+	if parts[1] == "" {
+		return TaskID{}, fmt.Errorf("missing node name in task ID: %s", taskID)
+	}
+
+	pid, err := strconv.ParseInt(parts[2], 16, 32)
+	if err != nil {
+		return TaskID{}, fmt.Errorf("error parsing task ID: %w", err)
+	}
+
+	pstart, err := strconv.ParseInt(parts[3], 16, 32)
+	if err != nil {
+		return TaskID{}, fmt.Errorf("error parsing pstart in task ID: %q: %w", taskID, err)
+	}
+
+	stime, err := strconv.ParseInt(parts[4], 16, 32)
+	if err != nil {
+		return TaskID{}, fmt.Errorf("error parsing start time in task ID: %q: %w", taskID, err)
+	}
+
+	if parts[5] == "" {
+		return TaskID{}, fmt.Errorf("missing task type in task ID: %q", taskID)
+	}
+
+	if parts[7] == "" {
+		return TaskID{}, fmt.Errorf("missing user in task ID: %q", taskID)
+	}
+
+	return TaskID{
+		NodeName:  parts[1],
+		PID:       int(pid),
+		PStart:    int(pstart),
+		StartTime: time.Unix(stime, 0),
+		Type:      parts[5],
+		ID:        parts[6],
+		User:      parts[7],
+	}, nil
 }

--- a/proxmox/nodes/tasks/tasks_types_test.go
+++ b/proxmox/nodes/tasks/tasks_types_test.go
@@ -19,6 +19,7 @@ func TestParseTaskID(t *testing.T) {
 
 	stime, err := time.Parse(time.RFC3339, "2023-08-30T21:28:16-04:00")
 	require.NoError(t, err)
+
 	stime = stime.UTC()
 
 	tests := []struct {

--- a/proxmox/nodes/tasks/tasks_types_test.go
+++ b/proxmox/nodes/tasks/tasks_types_test.go
@@ -1,0 +1,90 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+package tasks
+
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestParseTaskID(t *testing.T) {
+	t.Parallel()
+
+	stime, err := time.Parse(time.RFC3339, "2023-08-30T21:28:16-04:00")
+	require.NoError(t, err)
+
+	tests := []struct {
+		name    string
+		taskID  string
+		want    TaskID
+		wantErr bool
+	}{
+		{
+			name:   "imgcopy task",
+			taskID: "UPID:pve:00061CB3:010BA69C:64EFECB0:imgcopy::root@pam:",
+			want: TaskID{
+				NodeName:  "pve",
+				PID:       400563,
+				PStart:    17540764,
+				StartTime: stime,
+				Type:      "imgcopy",
+				ID:        "",
+				User:      "root@pam",
+			},
+		},
+		{
+			name:   "qmcreate task",
+			taskID: "UPID:pve:00061CB3:010BA69C:64EFECB0:qmcreate:101:root@pam:",
+			want: TaskID{
+				NodeName:  "pve",
+				PID:       400563,
+				PStart:    17540764,
+				StartTime: stime,
+				Type:      "qmcreate",
+				ID:        "101",
+				User:      "root@pam",
+			},
+		},
+		{
+			name:    "missing node",
+			taskID:  "UPID::00061CB3:010BA69C:64EFECB0:qmcreate:101:root@pam:",
+			wantErr: true,
+		},
+		{
+			name:    "wrong ID format",
+			taskID:  "blah",
+			wantErr: true,
+		},
+		{
+			name:    "missing pid",
+			taskID:  "UPID:pve::010BA69C:64EFECB0:qmcreate:101:root@pam:",
+			wantErr: true,
+		},
+		{
+			name:    "missing parts",
+			taskID:  "UPID:pve:00061CB3:010BA69C:64EFECB0::root@pam:",
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		tt := tt // capture range variable
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			got, err := ParseTaskID(tt.taskID)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("ParseTaskID() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("ParseTaskID() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}

--- a/proxmox/nodes/tasks/tasks_types_test.go
+++ b/proxmox/nodes/tasks/tasks_types_test.go
@@ -19,6 +19,7 @@ func TestParseTaskID(t *testing.T) {
 
 	stime, err := time.Parse(time.RFC3339, "2023-08-30T21:28:16-04:00")
 	require.NoError(t, err)
+	stime = stime.UTC()
 
 	tests := []struct {
 		name    string

